frontend/src/pages/pipelines/global/runs/RestoreRunWithArchivedExperimentModal.tsx

```diff
@@ -0,0 +1,209 @@
+import RunsWithArchivedExperimentTable from './RunsWithArchivedExperimentTable';
+import { PipelineRunTabTitle } from './types';
+
+type RestoreRunWithArchivedExperimentModalProps = {
+  selectedRuns: PipelineRunKF[];
+  archivedExperiments: ExperimentKF[];
+  onClose: (restored: boolean) => void;
+};
+
+const RestoreRunWithArchivedExperimentModal: React.FC<
+  RestoreRunWithArchivedExperimentModalProps
+> = ({ selectedRuns, archivedExperiments, onClose }) => {
+  const [isExpanded, setIsExpanded] = React.useState<boolean>(true);
+  const [isRetrying, setIsRetrying] = React.useState<boolean>(false);
+  const isSingleRestoring = selectedRuns.length === 1;
+
+  const {
+    onBeforeClose,
+    failedRuns,
+    failedExperiments,
+    error,
+    isSubmitting,
+    setExperimentStatuses,
+    setRunStatuses,
+    setIsSubmitting,
+    setError,
+    runStatuses,
+    experimentStatuses,
+    abortSignal,
+  } = useRestoreStatuses({
+    onClose,
+    isSingleRestoring,
+    selectedRuns,
+    archivedExperiments,
+  });
+
+  React.useEffect(() => {
+    if (failedExperiments.length > 0 || failedRuns.length > 0) {
+      setIsRetrying(true);
+    } else {
+      setIsRetrying(false);
+    }
+  }, [failedExperiments.length, failedRuns.length]);
+
+  const { api } = usePipelinesAPI();
+
+  const restoreProcess = React.useCallback(
+    async (runs: PipelineRunKF[], experiments: ExperimentKF[]) => {
+      const [, , experimentResults] = await allSettledPromises(
+        experiments.map((exp) =>
+          api.unarchiveExperiment({ signal: abortSignal }, exp.experiment_id),
+        ),
+      );
+      const [, , runResults] = await allSettledPromises(
+        runs.map((run) => api.unarchivePipelineRun({ signal: abortSignal }, run.run_id)),
+      );
+      return { experimentResults, runResults };
+    },
+    [abortSignal, api],
+  );
+
+  const onSubmit = React.useCallback(async () => {
+    setIsSubmitting(true);
+    setError(undefined);
+    setExperimentStatuses([]);
+    setRunStatuses([]);
+    const experimentsToRestore = archivedExperiments;
+
+    const runsToRestore = selectedRuns;
+
+    const { experimentResults, runResults } = await restoreProcess(
+      runsToRestore,
+      experimentsToRestore,
+    );
+    onBeforeClose(
+      true,
+      experimentResults.map((result) => (result.status === 'fulfilled' ? true : result.reason)),
+      runResults.map((result) => (result.status === 'fulfilled' ? true : result.reason)),
+    );
+  }, [
+    archivedExperiments,
+    onBeforeClose,
+    restoreProcess,
+    selectedRuns,
+    setError,
+    setExperimentStatuses,
+    setIsSubmitting,
+    setRunStatuses,
+  ]);
+
+  return (
+    <Modal
+      title={`Restore ${isSingleRestoring ? 'run' : `${selectedRuns.length} runs`}`}
+      titleIconVariant="warning"
+      isOpen
+      data-testid="restore-run-modal"
+      onClose={() => onBeforeClose(false)}
+      actions={[
```

Comment:
I'd suggest you use `DashboardModalFooter` (You can refer to many other places) to deal with the submit/cancel actions and error alert. It can fix the error alert always at the bottom of the modal instead of making it at the end of the modal body. So it won't be scrolled together with the modal if the body is too long.